sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -318,9 +316,8 @@ object Cast extends QueryErrorsBase {
     case _ if from == to => true
     case (NullType, _) => true
     case (_: NumericType, _: NumericType) => true
-    case (_: AtomicType, CharType(_) | VarcharType(_)) => false
-    case (_: CalendarIntervalType, CharType(_) | VarcharType(_)) => false
     case (_: AtomicType, _: StringType) => true
+    case (_: CalendarIntervalType, CharType(_) | VarcharType(_)) => false
```

Comment:
why only forbid `CalendarIntervalType` but not `AtomicType`?

Reply:
If we want to avoid breaking change like https://github.com/apache/spark/pull/49340/files#diff-928d574dedbca234036aa1f8b0098efdf914b74d2d6071cb439ab3d2025a25c3R703 , then we should allow both of them.